iwyu_ast_util.cc

```diff
@@ -270,12 +272,81 @@ bool IsNodeInsideCXXMethodBody(const ASTNode* ast_node) {
   return false;
 }
 
+// This is subtle and a little complicated.
+//
+// In order to override a method, we must derive from a base class, and to do
+// that we must include the base class header. So we can use that as an include
+// optimization; there's no reason for derived class headers to include or
+// declare all the same types as their base class header for the purpose of
+// spelling out the override. We already know all those types are available
+// because the base class header _is_ included.
+
+// There's no canonical place in AST traversal to detect if a use is part of an
+// overridden method signature, VisitFunctionDecl reports full uses of return
+// type and 'autocast' parameters. All other uses are reported as descendant AST
+// nodes are visited (such as exception specifications, pointer and reference
+// type parameters, etc.)
+//
+// Instead, whenever a use is reported we walk back up the tree and see if
+// we're part of an override signature, which can then be ignored.
+//
+// Return true if the given ast_node is inside the signature of a C++ method
+// that overrides a virtual method in a base class.
+static bool IsNodeInsideCXXMethodOverrideSignature(const ASTNode* ast_node) {
+  // This analysis only makes sense for the signature, not the body.
+  // But the IsNodeInsideCXXMethodBody check is basically as expensive as this
+  // entire function, so let's not do it again.
+  // CHECK_(!IsNodeInsideCXXMethodBody(ast_node));
+
+  // Walk up and find a containing CXXMethodDecl, and make sure it overrides a
+  // base method.
+  const auto* method = ast_node->GetAncestorOfType<CXXMethodDecl>();
+  if (!method || !method->size_overridden_methods()) {
+    return false;
+  }
+
+  // Now we know this node is part of a C++ method signature overriding a base.
+
+  if (HasCovariantReturnType(method)) {
+    // Covariant return types require that the full return type is available,
+    // so don't special-case the return type for functions with covariance.
+  } else {
```

Comment:
I find this a bit confusing. Why not invert the condition and explain what we do for non-covariant return types?